virtool/uploads/data.py

```diff
@@ -0,0 +1,199 @@
+
+import virtool.utils
+from virtool.api.response import NotFound
+from virtool.data.piece import DataLayerPiece
+
+
+from virtool.mongo.core import DB
+from virtool.mongo.transforms import apply_transforms
+from virtool.pg.base import Base
+
+from virtool.uploads.models import Upload
+from virtool.users.db import AttachUserTransform
+from virtool.utils import run_in_thread, rm
+
+logger = getLogger(__name__)
+
+
+class UploadsData(DataLayerPiece):
+    def __init__(self, config, db, pg):
+        self._config = config
+        self._db: DB = db
+        self._pg = pg
+
+    async def find(
+        self, user: str = None, upload_type: str = None, ready: bool = None
+    ) -> List[UploadMinimal]:
+        """
+        Find and filter uploads.
+        """
+
+        filters = [Upload.removed == False]
+        uploads = []
+
+        async with AsyncSession(self._pg) as session:
+            if user:
+                filters.append(Upload.user == user)
+
+            if upload_type:
+                filters.append(Upload.type == upload_type)
+
+            if ready is not None:
+                filters.append(Upload.ready == ready)
+
+            results = await session.execute(select(Upload).filter(*filters))
+
+        for result in results.unique().scalars().all():
+            uploads.append(result.to_dict())
+
+        uploads = await apply_transforms(uploads, [AttachUserTransform(self._db)])
+
+        return [UploadMinimal(**upload) for upload in uploads]
+
+    async def create(
+        self,
+        name: str,
+        upload_type: str,
+        reserved: bool = False,
+        user: Optional[str] = None,
+    ) -> Upload:
+        """
+        Create an upload.
+        """
+
+        async with AsyncSession(self._pg) as session:
+            upload = Upload(
+                created_at=virtool.utils.timestamp(),
+                name=name,
+                ready=False,
+                removed=False,
+                reserved=reserved,
+                type=upload_type,
+                user=user,
+            )
+
+            session.add(upload)
+
+            await session.flush()
+
+            upload.name_on_disk = f"{upload.id}-{upload.name}"
+
+            upload = upload.to_dict()
+
+            await session.commit()
+
+        return Upload(**upload)
+
+    async def get(self, upload_id: int) -> Optional[Upload]:
+        """
+        Get a single upload by its ID.
+
+        :param upload_id: the upload's ID
+        :return: the upload
+        """
+        async with AsyncSession(self._pg) as session:
+            upload = (
+                await session.execute(
+                    select(Upload).filter_by(id=upload_id, removed=False)
+                )
+            ).scalar()
+
+            if not upload:
+                return None
+
+        return upload
+
+    async def delete(self, upload_id: int) -> Upload:
+        """
+        "Delete" a row in the `uploads` table and remove it from the local disk.
```

Comment:
The data layer is meant to abstract away the underlying services like MongoDB, SQL, Redis, storage media, and more. This means the layer looks them same even if implementation is changing behind the scenes.

In this light, the docstring doesn't make sense. Come up with a docstring that says what the method does without talking about the underlying implementation.